core/shared/src/main/scala/zio/Chunk.scala

```diff
@@ -1213,6 +1214,37 @@ object Chunk extends ChunkFactory with ChunkPlatformSpecific {
   def succeed[A](a: A): Chunk[A] =
     single(a)
 
+  /**
+   * Constructs a `Chunk` by repeatedly applying the function `f` as long as it
+   * returns `Some`.
+   */
+  def unfold[S, A](s: S)(f: S => Option[(A, S)]): Chunk[A] = {
+
+    @tailrec
+    def go(s: S, builder: ChunkBuilder[A]): Chunk[A] =
+      f(s) match {
+        case Some((a, s)) => go(s, builder += a)
+        case None         => builder.result()
+      }
+
+    go(s, ChunkBuilder.make[A]())
+  }
+
+  /**
+   * Constructs a `Chunk` by repeatedly applying the effectual function `f` as
+   * long as it returns `Some`.
+   */
+  def unfoldM[R, E, A, S](s: S)(f: S => ZIO[R, E, Option[(A, S)]]): ZIO[R, E, Chunk[A]] = {
+
+    def go(s: S, builder: ChunkBuilder[A]): ZIO[R, E, Chunk[A]] =
+      f(s).flatMap {
+        case Some((a, s)) => go(s, builder += a)
+        case None         => ZIO.succeedNow(builder.result())
+      }
+
+    go(s, ChunkBuilder.make[A]())
```

Comment:
I think you need to use `effectSuspend` here, because the ChunkBuilder will be shared between all invocations of the ZIO returned from `unfoldM`.

Reply:
Good point!